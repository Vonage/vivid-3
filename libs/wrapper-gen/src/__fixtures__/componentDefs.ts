import { ComponentDef } from '../metadata/ComponentDef';

export const minimalComponent: ComponentDef = {
	name: 'example-component',
	className: 'ExampleComponent',
	wrappedClassName: 'VExampleComponent',
	vividModulePath:
		'libs/components/src/lib/example-component/example-component.ts',
	registerFunctionName: 'registerExampleComponent',
	attributes: [],
	events: [],
	vueModels: [],
	methods: [],
	slots: [],
	localTypeDefs: {},
};

export const exampleComponent: ComponentDef = {
	name: 'example-component',
	className: 'ExampleComponent',
	wrappedClassName: 'VExampleComponent',
	vividModulePath:
		'libs/components/src/lib/example-component/example-component.ts',
	registerFunctionName: 'registerExampleComponent',
	description: 'This is an example component',
	attributes: [
		{
			name: 'example-attribute',
			description: 'This is an example attribute',
			forwardTo: {
				type: 'attribute',
				name: 'example-attribute',
			},
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'type-union',
			forwardTo: {
				type: 'attribute',
				name: 'type-union',
			},
			type: [
				{ text: 'string', vuePropType: 'String' },
				{ text: 'number', vuePropType: 'Number' },
			],
		},
		{
			name: 'imported-type',
			forwardTo: {
				type: 'property',
				name: 'importedType',
			},
			type: [
				{
					text: 'Imported',
					vuePropType: 'Object',
					importFromModule: 'module',
				},
			],
		},
		{
			name: 'value',
			forwardTo: {
				type: 'attribute',
				name: 'value',
			},
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'start',
			forwardTo: {
				type: 'attribute',
				name: 'start',
			},
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'forced-dom-prop',
			forwardTo: {
				type: 'property',
				name: 'forcedDomProp',
			},
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'boolean-attribute',
			forwardTo: {
				type: 'attribute',
				name: 'boolean-attribute',
				boolean: true,
			},
			type: [
				{
					text: 'boolean',
					vuePropType: 'Boolean',
				},
			],
		},
	],
	events: [
		{
			name: 'example-event',
			description: 'This is an example event',
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'input',
			type: [
				{
					text: 'Event',
					vuePropType: 'Event',
				},
			],
		},
		{
			name: 'input:start',
			type: [
				{
					text: 'CustomEvent<{a: string | number}>',
					vuePropType: 'CustomEvent',
				},
			],
		},
	],
	vueModels: [
		{
			name: 'modelValue',
			attributeName: 'value',
			eventNames: ['input'],
			valueMapping: '(event.target as any).value',
		},
		{
			name: 'sameEvent',
			attributeName: 'example-attribute',
			eventNames: ['input'],
			valueMapping: '(event.target as any).exampleAttribute',
		},
		{
			name: 'start',
			attributeName: 'start',
			eventNames: ['input:start'],
			valueMapping: '(event.target as any).start',
		},
	],
	methods: [
		{
			name: 'exampleMethod',
			description: 'This is an example method',
			args: [
				{
					name: 'exampleArg',
					type: [
						{
							text: 'string',
							vuePropType: 'String',
						},
					],
				},
			],
			returnType: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
	],
	slots: [
		{
			name: 'example-slot',
			description: 'This is an example slot',
		},
		{
			name: 'no-description',
		},
	],
	localTypeDefs: {},
};
