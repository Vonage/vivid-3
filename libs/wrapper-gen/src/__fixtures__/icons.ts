import { IconsManifest } from '@vonage/vwc-consts';

export const iconsManifestStub: IconsManifest = [
	{
		id: '1-icon-1',
		keyword: ['icon-1'],
		tag: ['style_weight_solid', 'style_color_single', 'category_general'],
	},
	{
		id: 'icon-2',
		keyword: ['icon-2'],
		tag: ['style_weight_regular', 'style_color_multi', 'category_general'],
	},
];
