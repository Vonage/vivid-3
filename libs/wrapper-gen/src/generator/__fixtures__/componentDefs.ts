import { ComponentDef } from '../ComponentDef';

export const minimalComponent: ComponentDef = {
	name: 'example-component',
	className: 'ExampleComponent',
	wrappedClassName: 'VExampleComponent',
	vividModulePath:
		'libs/components/src/lib/example-component/example-component.ts',
	registerFunctionName: 'registerExampleComponent',
	attributes: [],
	events: [],
	vueModels: [],
	methods: [],
	slots: [],
	localTypeDefs: {},
};

export const exampleComponent: ComponentDef = {
	name: 'example-component',
	className: 'ExampleComponent',
	wrappedClassName: 'VExampleComponent',
	vividModulePath:
		'libs/components/src/lib/example-component/example-component.ts',
	registerFunctionName: 'registerExampleComponent',
	description: 'This is an example component',
	attributes: [
		{
			name: 'example-attribute',
			description: 'This is an example attribute',
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'type-union',
			type: [
				{ text: 'string', vuePropType: 'String' },
				{ text: 'number', vuePropType: 'Number' },
			],
		},
		{
			name: 'imported-type',
			type: [
				{
					text: 'Imported',
					vuePropType: 'Object',
					importFromModule: 'module',
				},
			],
		},
		{
			name: 'value',
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'start',
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'forced-dom-prop',
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
			forceDomProp: true,
		},
	],
	events: [
		{
			name: 'example-event',
			description: 'This is an example event',
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'input',
			type: [
				{
					text: 'Event',
					vuePropType: 'Event',
				},
			],
		},
		{
			name: 'input:start',
			type: [
				{
					text: 'CustomEvent<{a: string | number}>',
					vuePropType: 'CustomEvent',
				},
			],
		},
	],
	vueModels: [
		{
			name: 'modelValue',
			attributeName: 'value',
			eventName: 'input',
			valueMapping: '(event.target as any).value',
		},
		{
			name: 'start',
			attributeName: 'start',
			eventName: 'input:start',
			valueMapping: '(event.target as any).start',
		},
	],
	methods: [
		{
			name: 'exampleMethod',
			description: 'This is an example method',
			args: [
				{
					name: 'exampleArg',
					type: [
						{
							text: 'string',
							vuePropType: 'String',
						},
					],
				},
			],
			returnType: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
	],
	slots: [
		{
			name: 'example-slot',
			description: 'This is an example slot',
		},
		{
			name: 'no-description',
		},
	],
	localTypeDefs: {},
};
