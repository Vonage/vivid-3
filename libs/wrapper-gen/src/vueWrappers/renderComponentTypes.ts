import { ComponentDef } from '../common/ComponentDef';
import { parseTypeStr } from '../common/types';
import { Import, importsForTypes, renderImports } from './imports';
import { getExportedClassName } from '../metadata/vividPackage';
import { kebabToCamel, kebabToPascal } from '../utils/casing';
import { wrappedComponentName } from './name';
import { getEventType } from './types';
import { renderJsDoc } from './jsDoc';
import { resolveVueModels } from './vueModels';

export const renderComponentTypes = (componentDef: ComponentDef) => {
	const { props, vueModelEvents } = resolveVueModels(componentDef);

	const typeImports: Import[] = [
		{
			name: getExportedClassName(componentDef.name),
			fromModule: '@vonage/vivid',
		},
	];

	// Import referenced types
	const referencesTypes = [
		...props.map((prop) => prop.type),
		...componentDef.events.map((event) => event.type),
	].flatMap(parseTypeStr);
	typeImports.push(...importsForTypes(referencesTypes));

	const propTypesSrc = props
		.flatMap((prop) => {
			const vueModel = componentDef.vueModels.find(
				(model) => model.propName === prop.name
			);

			return [
				prop,
				...(vueModel && vueModel.name !== prop.name
					? [
							{
								...prop,
								name: vueModel.name,
							},
					  ]
					: []),
			];
		})
		.map(({ name, description, type }) => {
			const propName = kebabToCamel(name);
			return `${renderJsDoc(description)}
        ${propName}?: ${type}`;
		})
		.join(',\n');

	const renderEventHandler = ({
		name,
		description,
	}: ComponentDef['events'][number]) =>
		`
				${renderJsDoc(description)}
				'on${kebabToPascal(name)}'?: (event: ${wrappedComponentName(
			componentDef
		)}Events['${name}']) => void`;

	const eventHandlersSrc = [...componentDef.events, ...vueModelEvents]
		.map(renderEventHandler)
		.join(',\n');

	const renderEventType =
		(isVueModelEvent: boolean) =>
		({ name, description, type }: ComponentDef['events'][number]) =>
			`
				${renderJsDoc(description)}
				'${name}': ${getEventType(
				type,
				getExportedClassName(componentDef.name),
				isVueModelEvent
			)}`;

	const eventTypesSrc = [
		...componentDef.events.map(renderEventType(false)),
		...vueModelEvents.map(renderEventType(true)),
	].join(',\n');

	return `
${renderImports(typeImports, true)}

export interface ${wrappedComponentName(componentDef)}Events {
	${eventTypesSrc}
}

export interface ${wrappedComponentName(componentDef)}Props {
	${propTypesSrc}
	${eventHandlersSrc}
}
`;
};
