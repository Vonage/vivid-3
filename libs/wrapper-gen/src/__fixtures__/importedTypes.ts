import { makeTypeResolver } from '../common/types';
import { iconTypeFromManifest } from '../common/icons';
import { iconsManifestStub } from './icons';

export const importedTypesResolverStub = makeTypeResolver({
	IconId: iconTypeFromManifest(iconsManifestStub),
});
