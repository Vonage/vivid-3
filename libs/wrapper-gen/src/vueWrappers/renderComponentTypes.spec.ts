import { exampleComponent } from '../__fixtures__/componentDefs';
import { renderComponentTypes } from './renderComponentTypes';

describe('renderComponentTypes', () => {
	it('should render a type definitions file', () => {
		expect(renderComponentTypes(exampleComponent)).toMatchSnapshot();
	});
});
