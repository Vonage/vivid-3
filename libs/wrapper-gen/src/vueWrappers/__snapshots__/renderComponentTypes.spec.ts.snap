// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renderComponentTypes > should render a type definitions file 1`] = `
"
import type { VwcExampleComponentElement } from '@vonage/vivid';
import type { IconId } from '../icons';

export interface VExampleComponentEvents {
	
				/**
 * This is an example event
 */
				'example-event': Event & {
			/**
			 * @deprecated Target may not refer to component in some cases. Use currentTarget instead.
			 */
			target: VwcExampleComponentElement,
			currentTarget: VwcExampleComponentElement
		},

				
				'input': Event & {
			/**
			 * @deprecated Target may not refer to component in some cases. Use currentTarget instead.
			 */
			target: VwcExampleComponentElement,
			currentTarget: VwcExampleComponentElement
		},

				
				'input:start': CustomEvent<{a: string | number}> & {
			/**
			 * @deprecated Target may not refer to component in some cases. Use currentTarget instead.
			 */
			target: VwcExampleComponentElement,
			currentTarget: VwcExampleComponentElement
		},

				/**
 * Fires when the modelValue value changes
 */
				'update:modelValue': string,

				/**
 * Fires when the sameEvent value changes
 */
				'update:sameEvent': string,

				/**
 * Fires when the start value changes
 */
				'update:start': string
}

export interface VExampleComponentProps {
	/**
 * This is an example prop
 */
        exampleProp?: string,
/**
 * This is an example prop
 */
        sameEvent?: string,

        typeUnion?: string | number,

        iconType?: IconId,

        value?: string,

        modelValue?: string,

        start?: string,

        forcedDomProp?: string | HTMLElement,

        booleanProp?: boolean
	
				/**
 * This is an example event
 */
				'onExampleEvent'?: (event: VExampleComponentEvents['example-event']) => void,

				
				'onInput'?: (event: VExampleComponentEvents['input']) => void,

				
				'onInput:start'?: (event: VExampleComponentEvents['input:start']) => void,

				/**
 * Fires when the modelValue value changes
 */
				'onUpdate:modelValue'?: (event: VExampleComponentEvents['update:modelValue']) => void,

				/**
 * Fires when the sameEvent value changes
 */
				'onUpdate:sameEvent'?: (event: VExampleComponentEvents['update:sameEvent']) => void,

				/**
 * Fires when the start value changes
 */
				'onUpdate:start'?: (event: VExampleComponentEvents['update:start']) => void
}
"
`;
